### PR TITLE
[PF-2323] update cpu and gpu in instance in `updateAiNotebookInstance`

### PR DESCRIPTION
**Pull request by ginaygoogle**
This PR comes from [PF-2323](https://broadworkbench.atlassian.net/browse/PF-2323) for supporting to update cpu and show the cpu gpu information in return

The PR's scope is :
- modify the update ai notebook api with cpu gpu
- show the ai notebook information with cpu gpu by get/create/update

The PR code changes includes:
- API yaml file add `machineType` (cpu) and `acceleratorConfig` (gpu) in `GcpAiNotebookInstanceResource` as input
- API yaml file add `machineType` (cpu) and `acceleratorConfig` (gpu)  as output
- modify the input parameters in API controller and`controlledResourceService`
- create a new flight step for only updating CPU GPU
- create a new flight step for checking the instance is stopped before updating cpu gpu
- unit test/connected test/integration test

The manually testing:
**get ainotebook, return with the cpu gpu**:
Screenshot 2023-01-10 at 4 07 01 PM

**update ainotebook with cpu gpu, return the updated cpu gpu**:
Screenshot 2023-01-10 at 4 12 03 PM

[PF-2323]: https://broadworkbench.atlassian.net/browse/PF-2323?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: service/src/test/java/bio/terra/workspace/common/fixtures/ControlledResourceFixtures.java [hunk 1]
```diff
@@ -277,7 +268,7 @@ public static ApiGcpAiNotebookInstanceCreationParameters defaultNotebookCreation
     return new ApiGcpAiNotebookInstanceCreationParameters()
         .instanceId("default-instance-id")
         .location("us-east1-b")
-        .machineType("e2-standard-2")
+        .machineType("n1-standard-2")
```
Comment (ginaygoogle) [inline service/src/test/java/bio/terra/workspace/common/fixtures/ControlledResourceFixtures.java:271]: From the https://cloud.google.com/compute/docs/general-purpose-machines we know the e2 type does not support Gpu.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java [hunk 2]
```diff
@@ -120,14 +121,6 @@ public class ControlledResourceServiceTest extends BaseConnectedTest {
   /** The default GCP location to create notebooks for this test. */
   private static final String DEFAULT_NOTEBOOK_LOCATION = "us-east1-b";
 
-  private static final String DEST_DATASET_NAME = TestUtils.appendRandomNumber("dest_dataset_name");
```
Comment (ginaygoogle) [inline service/src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java:123]: In this file, I delete many did not used functions and variables, also remove some did not used exceptions to help the file clean and efficient.

**Comment by ginaygoogle** [comment]
After we talk after the standup. The code still need to modify the response parameters. So the PR in still in progress. Not ready to review.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java [hunk 3]
```diff
@@ -57,15 +77,28 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
       }
       sanitizedMetadata.put(entrySet.getKey(), entrySet.getValue());
     }
+    String machineType = inputMap.get(MACHINE_TYPE, String.class);
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java:80]: do you want to use the UPDATE_MACHINE_TYPE instead?

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java [hunk 4]
```diff
@@ -109,4 +148,115 @@ private StepResult updateAiNotebook(Map<String, String> metadataToUpdate, String
...
+          .instances()
+          .setMachineResources(projectId, location, instanceId, setMachineResourcesRequest)
+          .execute();
+    }
+
+    // Wait for the instance to finish updating
+    int retryCount = 5, retryWaitSeconds = 2;
+
+    for (int i = 0; i < retryCount; i++) {
+      Instance instanceInfo =
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java:219]: can you add a retry in GcpUtils.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java [hunk 5]
```diff
@@ -109,4 +148,115 @@ private StepResult updateAiNotebook(Map<String, String> metadataToUpdate, String
...
+   * example in https://cloud.google.com/compute/docs/reference/rest/v1/instances/get
+   */
+  public static Compute createComputeService() throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+
+    GoogleCredential credential = GoogleCredential.getApplicationDefault();
+    if (credential.createScopedRequired()) {
+      credential =
+          credential.createScoped(Arrays.asList("https://www.googleapis.com/auth/cloud-platform"));
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java:251]: can you extract this to CrlService?

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java:251]: similar to https://github.com/DataBiosphere/terra-workspace-manager/blob/3bd1f03cd912ea8fe93e5678a7b97876df71f1a4/service/src/main/java/bio/terra/workspace/service/crl/CrlService.java#L237.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookCpuGpuStep.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/** {@link Step} to update machine type(cpu) or accelerator config(gpu) for ai notebook instance. */
+public class UpdateAiNotebookCpuGpuStep implements Step {
+  private final ControlledAiNotebookInstanceResource resource;
+  private final CrlService crlService;
+  private final GcpCloudContextService cloudContextService;
+
+  private final Logger logger = LoggerFactory.getLogger(UpdateAiNotebookCpuGpuStep.class);
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookCpuGpuStep.java:43]: nit private static final

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookCpuGpuStep.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+          new InstancesSetMachineResourcesRequest();
+      setMachineResourcesRequest.setGuestAccelerators(List.of(acceleratorConfig));
+      computeService
+          .instances()
+          .setMachineResources(projectId, location, instanceId, setMachineResourcesRequest)
+          .execute();
+    }
+
+    // Wait for the instance to finish updating
+    int retryCount = 5, retryWaitSeconds = 2;
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookCpuGpuStep.java:167]: use the gcpUtils method once you extract it.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java [hunk 8]
```diff
@@ -109,4 +148,115 @@ private StepResult updateAiNotebook(Map<String, String> metadataToUpdate, String
     }
     return StepResult.getStepResultSuccess();
   }
+
+  private void updateAiNotebookCpuGpu(
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java:152]: revert the change in this file since you have an updateGpuCpuStep below?

File: service/src/test/java/bio/terra/workspace/common/GcpCloudUtils.java [hunk 9]
```diff
@@ -193,6 +202,26 @@ private static Storage getGcpStorageClient(GoogleCredentials userCredential, Str
         .build()
         .getService();
   }
+
+  /**
+   * Directly calling the gcp api to get/update instance, requires the createComputeService, see the
+   * example in https://cloud.google.com/compute/docs/reference/rest/v1/instances/get
+   */
+  public static Compute createComputeService() throws IOException, GeneralSecurityException {
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/common/GcpCloudUtils.java:210]: same here. if we have the shared method in CrlService, we don't need this one.

File: service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java [hunk 10]
```diff
@@ -51,6 +60,24 @@ public void setUp() {
     workspaceUuid = createWorkspaceWithGcpContext(workspaceDao);
   }
 
+  @BeforeEach
+  public void setUpMocks() throws IOException {
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java:64]: rm? I don't see any code changes in ResourceDao.

**Comment by ginaygoogle** [comment]
Still debugging now. Not ready for reviewing at this moment. 
I will let reviewer know when it done.

File: service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java (hunk 10 above)
Comment (ginaygoogle) [reply service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java:64]: Hi Yu. We need that because when we get the AinotebookInstance, they will get the cpu and gpu information of the instance. the instances.get API needs a mock here. See the screenshot when i remove the mock.
https://screenshot.googleplex.com/4wRToMauJrMq3os

I also feel a little bit strange but seems like this is the easiest way. let me know if you have better approaches.

Comment (yuhuyoyo) [reply service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java:64]: oic the handler.

@beforeAll.

Comment (ginaygoogle) [reply service/src/test/java/bio/terra/workspace/db/ResourceDaoTest.java:64]: I also tried to use the BeforeAll, but also fails like the screenshot. (Sorry, I should mentioned it in the last comment, I forget...) It seems like the mock will expire after each unitTest finish.

File: openapi/src/resources/gcp_ai_notebook_instance.yaml [hunk 11]
```diff
@@ -20,10 +20,15 @@ components:
     GcpAiNotebookInstanceResource:
       type: object
       description: Description of a AI Notebook Instance resource.
-      required: [metadata, attributes]
+      required: [metadata, attributes, machineType]
```
Comment (ginaygoogle) [inline openapi/src/resources/gcp_ai_notebook_instance.yaml:23]: The machineType could not be null. but acceleratorConfig could be null.
